x-pack/plugins/observability_solution/observability_ai_assistant/public/components/insight/insight.tsx

```diff
@@ -192,17 +193,56 @@ function PromptEdit({
 }
 
 export interface InsightProps {
-  messages: Message[];
+  messages: Message[] | (() => Promise<Message[] | undefined>);
   title: string;
   dataTestSubj?: string;
 }
 
-export function Insight({ messages, title, dataTestSubj }: InsightProps) {
-  const [initialMessages, setInitialMessages] = useState(messages);
+enum FETCH_STATUS {
+  LOADING = 'loading',
+  SUCCESS = 'success',
+  FAILURE = 'failure',
+  NOT_INITIATED = 'not_initiated',
+}
+
+export function Insight({
+  messages: initialMessagesOrCallback,
+  title,
+  dataTestSubj,
+}: InsightProps) {
+  const [messages, setMessages] = useState<Message[]>([]);
   const [isEditingPrompt, setEditingPrompt] = useState(false);
   const [isInsightOpen, setInsightOpen] = useState(false);
   const [hasOpened, setHasOpened] = useState(false);
   const [isPromptUpdated, setIsPromptUpdated] = useState(false);
+  const [messagesFetchStatus, setMessagesFetchStatus] = useState<FETCH_STATUS>(
+    FETCH_STATUS.NOT_INITIATED
+  );
+
+  const updateInitialMessages = useCallback(() => {
+    if (isArray(initialMessagesOrCallback)) {
+      setMessages(initialMessagesOrCallback);
```

Comment:
does this clean up if we have a single state (e.g. `{ status, messages }`)?

Reply:
Do you suggest changing the signature from:
```
messages: Message[] | (() => Promise<Message[] | undefined>);
```

to
```
messages: Message | (() => Promise<Message[] | undefined>);
```

?

Reply:
no, I think my comment ended up on the wrong line. What I mean is that you have two `useState` calls - one for `messages` and one for `messagesFetchStatus` (or something similar). Maybe it's cleaner if you use one `useState` call with `{ status, messages }`?

Reply:
Updated in fb60fea8ad44382313a663e14090ecc7c28b2ef3